rllib/examples/multi_agent_different_spaces_for_agents.py

```diff
@@ -0,0 +1,141 @@
+"""
+Example showing how one can create a multi-agent env, in which the different agents
+have different observation and action spaces.
+These spaces do NOT necessarily have to be specified manually by the user. Instead,
+RLlib will try to automatically infer them from the env provided spaces dicts
+(agentID -> obs/act space) and the policy mapping fn (mapping agent IDs to policy IDs).
+
+---
+Run this example with defaults (using Tune):
+
+  $ python multi_agent_different_spaces_for_agents.py
+"""
+
+import argparse
+import gym
+import os
+
+import ray
+from ray import tune
+from ray.rllib.env.multi_agent_env import MultiAgentEnv
+
+
+class BasicMultiAgentMultiSpaces(MultiAgentEnv):
+    """A simple multi-agent example environment where agents have different spaces.
+
+    agent0: obs=(10,), act=Discrete(2)
+    agent1: obs=(20,), act=Discrete(3)
+
+    The logic of the env doesn't really matter for this example. The point of this env
+    is to show how one can use multi-agent envs, in which the different agents utilize
+    different obs- and action spaces.
+    """
+    def __init__(self, config=None):
+        self.agents = {"agent0", "agent1"}
+        self._agent_ids = set(self.agents)
+
+        self.dones = set()
+
+        # Provide full (preferred format) observation- and action-spaces as Dicts
+        # mapping agent IDs to the individual agents' spaces.
+        self._spaces_in_preferred_format = True
+        self.observation_space = gym.spaces.Dict(
+            {"agent0": gym.spaces.Box(low="-1", high=1, shape=(10,)),
+             "agent1": gym.spaces.Box(low="-1", high=1, shape=(20,))})
+        self.action_space = gym.spaces.Dict(
+            {"agent0": gym.spaces.Discrete(2), "agent1": gym.spaces.Discrete(3)})
+
+        super().__init__()
+
+    def reset(self):
+        self.dones = set()
+        return {i: self.observation_space[i].sample() for i in self.agents}
+
+    def step(self, action_dict):
+        obs, rew, done, info = {}, {}, {}, {}
+        for i, action in action_dict.items():
+            obs[i], rew[i], done[i], info[i] = self.observation_space[
```

Comment:
done

Reply:
@gjoliver When is it estimated that TF1 is no longer supported? IMO this will increae code readability a lot and makes it easier to build algorithms.

Reply:
> @gjoliver When is it estimated that TF1 is no longer supported? IMO this will increae code readability a lot and makes it easier to build algorithms.

yeah I think so too. We need to collect stats about how much of our user base is on TF1, TF2, etc. and decide / communicate a date.

Reply:
> 

Thanks for clarification. Is `ray train` using also both or only TF2?